File: algorithms/algorithms-cplusplus/leetcode/Question_0084_Largest_Rectangle_in_Histogram.cc
#include <vector>
#include <stack>

using namespace std;

class Solution {
 public:
  int largestRectangleArea(vector<int>& heights) {
    if (heights.empty()) {
      return 0;
    }
    stack<int> st;
    int res = 0;
    for (int i = 0; i <= heights.size(); i++) {
      int cur = (i == heights.size()) ? 0 : heights[i];
      while (!st.empty() && heights[st.top()] >= cur) {
        int h = heights[st.top()];
        st.pop();
        int w = st.empty() ? i : i - st.top() - 1;
        res = max(res, h * w);
      }
      st.push(i);
    }
    return res;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0084_Largest_Rectangle_in_Histogram.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.Stack;

public class Question_0084_Largest_Rectangle_in_Histogram {
    public int largestRectangleArea(int[] heights) {
        if (heights == null) {
            return 0;
        }
        Stack<Integer> stack = new Stack<>();
        int res = 0;
        for (int i = 0; i <= heights.length; i++) {
            int cur = (i == heights.length) ? -1 : heights[i];
            while (!stack.empty() && heights[stack.peek()] >= cur) {
                int h = heights[stack.pop()];
                int w = stack.empty() ? i : i - stack.peek() - 1;
                res = Math.max(res, h * w);
            }
            stack.push(i);
        }
        return res;
    }
}
